neuro_san/internals/messages/base_message_dictionary_converter.py

```diff
@@ -0,0 +1,168 @@
+            if src == "content" and isinstance(value, list):
+                if len(value) > 0:
+                    value = value[0].get("text", "")
+                else:
+                    value = None
+
+            if value is not None:
+                chat_message[dest] = value
+
+        return chat_message
+
+    def from_dict(self, obj_dict: Dict[str, Any]) -> BaseMessage:
+        """
+        The intended use for this method is for converting a neuro-san ChatMessage
+        dictionary into a langchain BaseMessage such that it can be understood by
+        a langchain agent within the context of a langchain-facing chat history.
+        This means that some special messages we might have inserted into a chat
+        history *might* need to be converted into something langchain agents will
+        understand.
+
+        :param obj_dict: A ChatMessage dictionary to convert into BaseMessage
+        :return: A BaseMessage that was converted from the input.
+                Can return None if conversion could not take place
+        """
+        chat_message: Dict[str, Any] = obj_dict
+        base_message: BaseMessage = None
+        if chat_message is None:
+            return base_message
+
+        content: str = chat_message.get("text")
+        chat_message_type: ChatMessageType = ChatMessageType.from_response_type(chat_message.get("type"))
+
+        if chat_message_type == ChatMessageType.SYSTEM:
+            base_message = SystemMessage(content=content)
+        elif chat_message_type == ChatMessageType.HUMAN:
+            base_message = HumanMessage(content=content)
+        elif chat_message_type == ChatMessageType.AI:
+            base_message = AIMessage(content=content)
+        elif chat_message_type == ChatMessageType.AGENT_TOOL_RESULT:
+            base_message = AgentToolResultMessage(content=content,
+                                                  tool_result_origin=chat_message.get("tool_result_origin"))
+
+        if self.langchain_only:
+            # Go no further
+            return base_message
+
+        if base_message is None:
+            # convert_to_base_message() is primarily for populating langchain chat history.
+            # so handle these other types differently.
+            if chat_message_type == ChatMessageType.AGENT:
+                base_message = AgentMessage(content=chat_message.get("text", ""),
+                                            structure=chat_message.get("structure"))
+            elif chat_message_type == ChatMessageType.AGENT_FRAMEWORK:
+                # Don't bother passing on chat_context
+                base_message = AgentFrameworkMessage(content=chat_message.get("text", ""))
+
+        # Any other message type we do not want to send to any agent as chat history.
+
+        return base_message
+
+    @staticmethod
+    def is_relevant_to_chat_history(base_message: BaseMessage) -> bool:
```

Comment:
This implementation comes from MessageUtils.is_relevant_to_chat_history()
In the end, the intent is to keep all the BaseMessage <-> ChatMessage dictionary policy logic in one place, and this guy comes along for the ride.